### DISCUSSION TITLE
Implementing "Clock" In Lualine

### DISCUSSION THREAD

**Discussion by harrisoncramer** [nvim-lualine/lualine.nvim#493] [Ideas]
Hello All,

First off, thank you for the amazing plugin. It's exactly what I needed from a status line plugin.

I'm curious, is it possible to implement a function that will update? I'm currently using this Lua expression to get the current time:

`os.date("%I:%M:%S", os.time())`

However, this only updates when I type (which is odd) How might you implement this so that it's updated every second? Thanks again!

**Comment by shadmansaleh** [comment]
Lualine itself doesn't trigger any statusline redraw. Instead when
neovim tries to redraw statusline lualine provides what to draw.
Neovim only redraws statusline on certain events (as optimization). If you don't like the default
behavior, you can trigger statusline redraw with `:redrawstatus`  You can set a timer to update statusline every
1000ms like following this.

```lua
if _G.Statusline_timer == nil then
  _G.Statusline_timer = vim.loop.new_timer()
else
  _G.Statusline_timer:stop()
end
_G.Statusline_timer:start(0,             -- never timeout
                          1000,          -- repeat every 1000 ms
                          vim.schedule_wrap(function() -- updater function
                                              vim.api.nvim_command('redrawstatus')
                                            end))
```

**Reply by harrisoncramer** [reply]
That works! Thank you so much! For others interested, here is my full lualine config, with a clock:

```
require('lualine').setup({
    options = {theme = 'gruvbox-material'},
    sections = {
        lualine_a = {'branch'},
        lualine_b = {'diagnostics', 'diff'},
        lualine_c = {"vim.fn.expand('%')"},
        lualine_x = {'filetype'},
        lualine_y = {'os.date("%I:%M:%S", os.time())'},
        lualine_z = {}
    }
})

-- Turn off lualine inside nvim-tree
vim.cmd [[
  au BufEnter,BufWinEnter,WinEnter,CmdwinEnter * if bufname('%') == "NvimTree" | set laststatus=0 | else | set laststatus=2 | endif
]]

-- Trigger rerender of status line every second for clock
if _G.Statusline_timer == nil then
    _G.Statusline_timer = vim.loop.new_timer()
else
    _G.Statusline_timer:stop()
end
_G.Statusline_timer:start(0, 1000, vim.schedule_wrap(
                              function() vim.api.nvim_command('redrawstatus') end))

```

**Reply by dmartzol** [reply]
Can this be used to display UTC time? I tend to need UTC to schedule cronjobs in servers

**Reply by shadmansaleh** [reply]
@dmartzol prepend the time pattern with `!` to see UTC time regardless of your timezone. Example
```lua
os.date('!%I:%M:%S')
```
If `!` is not given it'll show you time of your timezone.

**Reply by dmartzol** [reply]
This is awesome, thank you!

**Reply by DartMitai** [reply]
why do I get the time in UTC on the contrary? but I need localtime...
`'os.date("%H:%M")'`
![image](https://user-images.githubusercontent.com/104931116/217727167-9c4eea99-5748-42d2-bbee-fc5a6cf213fe.png)
![image](https://user-images.githubusercontent.com/104931116/217727304-44608f1a-7d62-4392-8354-b91a6f5773c9.png)

**Reply by DartMitai** [reply]
and so in order to get a normal time, I was prompted with this option
`'os.date("!%H:%M", os.time() + (60 * 60 * 5))'`
however, why he always output UTC remains unclear...
